test/unittest/layers/layers_golden_tests.cpp

```diff
@@ -0,0 +1,206 @@
+    allocate_inouts(context.getInputDimensions()),
+    allocate_inouts(context.getOutputDimensions()),
+    allocate_tensors(context.getTensorsSpec()),
+  };
+}
+
+static RunLayerContext prepareRunContext(const TensorPacks &packs) {
+  auto &[weights, ins, outs, tensors] = packs;
+  auto create_view = [](const auto &var_grads) {
+    using ptr_type_ = std::add_pointer_t<
+      typename std::decay_t<decltype(var_grads)>::value_type>;
+    std::vector<std::remove_cv_t<ptr_type_>> ret;
+    ret.reserve(var_grads.size());
+
+    for (auto &vg : var_grads) {
+      ret.push_back(const_cast<ptr_type_>(&vg));
+    }
+
+    return ret;
+  };
+
+  auto rc =
+    RunLayerContext("golden", 0.0f, create_view(weights), create_view(ins),
+                    create_view(outs), create_view(tensors));
+
+  auto num_outputs = rc.getNumOutputs();
+
+  for (unsigned i = 0; i < num_outputs; ++i) {
+    rc.getOutput(i).setRandUniform(); /// randomize output
+    rc.getIncomingDerivative(i).setValue(
+      2.0); /// incoming derivative is fixed to 2
+  }
+
+  return rc;
+}
+
+static void compareRunContext(RunLayerContext &rc, std::ifstream &file) {
+  file.seekg(0, std::ios::beg);
+  auto compare_tensors = [&file](unsigned length, auto tensor_getter,
+                                 auto pred) {
+    for (unsigned i = 0; i < length; ++i) {
+      if (!pred(i)) {
+        continue;
+      }
+      const auto &tensor = tensor_getter(i);
+      auto answer = tensor.clone();
+      sizeCheckedReadTensor(answer, file);
+      EXPECT_EQ(tensor, answer);
+    }
+  };
+
+  auto always = [](unsigned idx) { return true; };
+  auto only_trainable = [&rc](unsigned idx) {
+    return rc.weightHasGradient(idx);
+  };
+
+  compare_tensors(rc.getNumWeights(),
+                  [&rc](unsigned idx) { return rc.getWeight(idx); }, always);
+  compare_tensors(rc.getNumInputs(),
+                  [&rc](unsigned idx) { return rc.getInput(idx); }, always);
+  compare_tensors(rc.getNumOutputs(),
+                  [&rc](unsigned idx) { return rc.getOutput(idx); }, always);
+  compare_tensors(rc.getNumWeights(),
+                  [&rc](unsigned idx) { return rc.getWeightGrad(idx); },
+                  only_trainable);
+  compare_tensors(rc.getNumInputs(),
+                  [&rc](unsigned idx) { return rc.getOutgoingDerivative(idx); },
+                  always);
+}
+
+LayerGoldenTest::~LayerGoldenTest() {}
+
+void LayerGoldenTest::SetUp() {}
+
+void LayerGoldenTest::TearDown() {}
+
+TEST_P(LayerGoldenTest, run) {
+  auto f = std::get<0>(GetParam());
+  auto layer = f(std::get<1>(GetParam()));
+  auto golden_file = checkedOpenStream<std::ifstream>(
+    getGoldenPath(std::get<3>(GetParam())), std::ios::in | std::ios::binary);
+  auto &input_dims = std::get<2>(GetParam());
+
+  auto ic = createInitContext(layer.get(), input_dims);
+  auto tensors = prepareTensors(ic, golden_file);
+  auto rc = prepareRunContext(tensors);
+
+  layer->forwarding(rc, true);
```

Comment:
this is good for now, but consider running for at least 2 iterations in later PRs.